File: src/Coding_Interview_Patterns_ByteByteGo/ch03_linked_lists/LinkedListReversal.java
package Coding_Interview_Patterns_ByteByteGo.ch03_linked_lists;

import static org.junit.jupiter.api.Assertions.*;

import Coding_Interview_Patterns_ByteByteGo.common.ListNode;

/**
 * <a href="https://bytebytego.com/exercises/coding-patterns/linked-lists/linked-list-reversal">Linked List Reversal</a>
 */
public class LinkedListReversal {

  public static void main(String[] args) {
    assertEquals(ListNode.fromArray(new int[] {3,7,4,2,1}), linkedListReversal(ListNode.fromArray(new int[] {1,2,4,7,3})));
  }

  /**
   * Approach: Recursive <br>
   * Time Complexity: O(n) <br>
   * Space Complexity: O(n) <br>
   */
  public static ListNode linkedListReversal(ListNode head) {
    if (head == null || head.next == null) {
      return head;
    }
    ListNode newHead = linkedListReversal(head.next);
    head.next.next = head;
    head.next = null;
    return newHead;
  }

//  /**
//   * Approach: Iterative <br>
//   * Time Complexity: O(n) <br>
//   * Space Complexity: O(1) <br>
//   */
//  public static ListNode linkedListReversal(ListNode head) {
//    ListNode currNode = head;
//    ListNode prevNode = null;
//    while (currNode != null) {
//      ListNode nextNode = currNode.next;
//      currNode.next = prevNode;
//      prevNode = currNode;
//      currNode = nextNode;
//    }
//    return prevNode;
//  }
}

File: src/ByteByteGo/Coding_Interview_Patterns/ch03_linked_lists/python3/linked_list_reversal.py
"""
Definition of ListNode
"""


class ListNode:
  def __init__(self, val=None, next=None):
    self.val = val
    self.next = next


'''
Approach: Recursive
Time Complexity: O(n)
Space Complexity: O(n)
'''


def linked_list_reversal(head: ListNode) -> ListNode:
  if (not head) or (not head.next):
    return head
  new_head = linked_list_reversal(head.next)
  head.next.next = head
  head.next = None
  return new_head


'''
Approach: Iterative
Time Complexity: O(n)
Space Complexity: O(1)
'''


# def linked_list_reversal(head: ListNode) -> ListNode:
#   curr_node = head
#   prev_node = None
#   while curr_node:
#     next_node = curr_node.next
#     curr_node.next = prev_node
#     prev_node = curr_node
#     curr_node = next_node
#   return prev_node
